modules/ethereum/src/lib.rs

```diff
@@ -286,15 +286,31 @@ pub trait Storage {
 	fn scheduled_change(&self, hash: &H256) -> Option<ScheduledChange>;
 	/// Insert imported header.
 	fn insert_header(&mut self, header: HeaderToImport<Self::Submitter>);
-	/// Finalize given block and prune all headers with number < prune_end.
+	/// Finalize given block and prune schedules pruning of all headers
+	/// with number < prune_end.
+	///
 	/// The headers in the pruning range could be either finalized, or not.
 	/// It is the storage duty to ensure that unfinalized headers that have
 	/// scheduled changes won't be pruned until they or their competitors
 	/// are finalized.
-	fn finalize_headers(&mut self, finalized: Option<HeaderId>, prune_end: Option<u64>);
+	fn finalize_headers(&mut self, finalized: Option<HeaderId>, prune_end: u64);
+}
+
+/// Headers pruning strategy.
+pub trait PruningStrategy {
+	/// Return upper bound (exclusive) of headers pruning range.
+	///
+	/// Every value that is returned from this function, must be greater or equal to the
+	/// previous value. Otherwise it will be ignored (we can't revert pruning).
+	///
+	/// Module may prune both finalized and unfinalized blocks. But it can't give any
+	/// guarantees on when it will happen. Example: if some unfinalized block at height N
+	/// has scheduled validators set change, then the module won't prune any blocks with
+	/// number >= N even if strategy allows that.
+	fn pruning_upper_bound(best_number: u64, best_finalized_number: u64) -> u64;
```

Comment:
I like how `Storage` traits require an instance and allows to pass different values if needed. I feel like we should do the same here (we can require `Default` bound to simplify creation of that instance). That would probably make the test code a bit nicer, since passing the type explicitly would not be required.

```suggestion
	fn pruning_upper_bound(&mut self, best_number: u64, best_finalized_number: u64) -> u64;
```